File: pysrc/coordsys.cpp
#include "coordsys.h"
#include <pybind11/pybind11.h>
#include <pybind11/numpy.h>
#include <pybind11/operators.h>
#include <pybind11/eigen.h>
#include <tuple>
#include <iostream>

namespace py = pybind11;

namespace batoid {
    void pyExportCoordSys(py::module& m) {
        py::class_<CoordSys, std::shared_ptr<CoordSys>>(m, "CPPCoordSys")
            .def(py::init<>())
            .def(py::init<Vector3d>())
            .def(py::init<Matrix3d>())
            .def(py::init<Vector3d,Matrix3d>())
            .def_readonly("origin", &CoordSys::m_origin, "Global origin")
            .def_readonly("rot", &CoordSys::m_rot, "Unit vector rotation matrix")
            .def_property_readonly("xhat", &CoordSys::getXHat)
            .def_property_readonly("yhat", &CoordSys::getYHat)
            .def_property_readonly("zhat", &CoordSys::getZHat)
            .def("__repr__", &CoordSys::repr)
            .def("shiftGlobal", &CoordSys::shiftGlobal)
            .def("shiftLocal", &CoordSys::shiftLocal)
            .def("rotateGlobal", (CoordSys (CoordSys::*) (const Matrix3d&) const) &CoordSys::rotateGlobal)
            .def("rotateGlobal", (CoordSys (CoordSys::*) (const Matrix3d&, const Vector3d&, const CoordSys&) const) &CoordSys::rotateGlobal)
            .def("rotateLocal", (CoordSys (CoordSys::*) (const Matrix3d&) const) &CoordSys::rotateLocal)
            .def("rotateLocal", (CoordSys (CoordSys::*) (const Matrix3d&, const Vector3d&, const CoordSys&) const) &CoordSys::rotateLocal)
            .def(py::self == py::self)
            .def(py::self != py::self)
            .def(py::pickle(
                [](const CoordSys& cs) { return py::make_tuple(cs.m_origin, cs.m_rot); },
                [](py::tuple t) {
                    return CoordSys(
                        t[0].cast<Vector3d>(), t[1].cast<Matrix3d>()
                    );
                }
            ))
            .def("__hash__", [](const CoordSys& cs) {
                return py::hash(py::make_tuple(
                    "CPPCoordSys",
                    py::tuple(py::cast(cs.m_origin)),
                    py::tuple(py::cast(cs.m_rot).attr("ravel")())
                ));
            });
    }
}

File: batoid/coordSys.py
import numpy as np


def RotX(th):
    sth, cth = np.sin(th), np.cos(th)
    return np.array([[1,0,0],[0,cth,-sth],[0,sth,cth]])


def RotY(th):
    sth, cth = np.sin(th), np.cos(th)
    return np.array([[cth,0,sth],[0,1,0],[-sth,0,cth]])


def RotZ(th):
    sth, cth = np.sin(th), np.cos(th)
    return np.array([[cth,-sth,0],[sth,cth,0],[0,0,1]])


class CoordSys:
    """A coordinate system against which to measure surfaces or rays.

    Coordinate systems consist of an origin and a rotation.  The ``origin``
    attribute specifies where in 3D space the current coordinate system's
    origin lands in the global coordinate system.  The rotation ``rot``
    specifies the 3D rotation matrix to apply to the global coordinate axes to
    yield the axes of the this coordinate system.

    Parameters
    ----------
    origin : ndarray of float, shape (3,)
        Origin of coordinate system in global coordinates.
    rot : ndarray of float, shape (3, 3)
        Rotation matrix taking global axes into current system axes.
    """
    def __init__(self, origin=None, rot=None):
        if origin is None:
            origin = np.zeros(3, dtype=float)
        if rot is None:
            rot = np.eye(3, dtype=float)
        self.origin = np.array(origin)
        self.rot = np.array(rot)

    @property
    def xhat(self):
        """ndarray of float, shape (3,): Orientation of local x vector in
        global coordinates.
        """
        return self.rot[:, 0]

    @property
    def yhat(self):
        """ndarray of float, shape (3,): Orientation of local y vector in
        global coordinates.
        """
        return self.rot[:, 1]

    @property
    def zhat(self):
        """ndarray of float, shape (3,): Orientation of local z vector in
        global coordinates.
        """
        return self.rot[:, 2]

    def shiftGlobal(self, dr):
        """Return new CoordSys with origin shifted along global axes.

        Parameters
        ----------
        dr : ndarray of float, shape (3,)
            Amount to shift in meters.

        Returns
        -------
        CoordSys
        """
        return CoordSys(self.origin+dr, self.rot)

    def shiftLocal(self, dr):
        """Return new CoordSys with origin shifted along local axes.

        Parameters
        ----------
        dr : ndarray of float, shape (3,)
            Amount to shift in meters.

        Returns
        -------
        CoordSys
        """
        # Rotate the shift into global coordinates, then do the shift globally
        return self.shiftGlobal(self.rot@dr)

    def rotateGlobal(self, rot, rotCenter=(0,0,0), coordSys=None):
        """Return new CoordSys rotated with respect to global axes.

        Parameters
        ----------
        rot : ndarray of float, shape (3, 3)
            Rotation matrix to apply.
        rotCenter : ndarray of float, shape (3,)
            Point about which to rotate.
        coordSys : CoordSys
            Coordinate system in which rotCenter is specified.

        Returns
        -------
        CoordSys
        """
        if coordSys is None:
            coordSys = CoordSys()
        # Find rot center in global coordinates
        globalRotCenter = coordSys.rot@rotCenter + coordSys.origin
        # Then rotate about this center
        return CoordSys(
            rot@(self.origin-globalRotCenter)+globalRotCenter,
            rot@self.rot
        )

    def rotateLocal(self, rot, rotCenter=(0,0,0), coordSys=None):
        """Return new CoordSys rotated with respect to local axes.

        Parameters
        ----------
        rot : ndarray of float, shape (3, 3)
            Rotation matrix to apply.
        rotCenter : ndarray of float, shape (3,)
            Point about which to rotate.
        coordSys : CoordSys
            Coordinate system in which rotCenter is specified.

        Returns
        -------
        CoordSys
        """
        if coordSys is None:
            coordSys = self
        # Find rot center in global coordinates
        globalRotCenter = coordSys.rot@rotCenter + coordSys.origin
        # first rotate rot into global coords: (self.rot rot self.rot.T),
        # then apply that: (self.rot rot self.rot.T) self.rot = self.rot rot
        rTmp = self.rot@rot
        return CoordSys(
            rTmp@self.rot.T@(self.origin-globalRotCenter)+globalRotCenter,
            rTmp
        )

    def __getstate__(self):
        return self.origin, self.rot

    def __setstate__(self, d):
        self.origin, self.rot = d

    def __eq__(self, rhs):
        if not isinstance(rhs, CoordSys): return False
        return (
            np.array_equal(self.origin, rhs.origin) and
            np.array_equal(self.rot, rhs.rot)
        )

    def __ne__(self, rhs):
        return not (self == rhs)

    def __hash__(self):
        return hash((
            "batoid.CoordSys",
            tuple(self.origin.tolist()),
            tuple(self.rot.ravel().tolist())
        ))

    def copy(self):
        return CoordSys(self.origin, self.rot)

    def __repr__(self):
        rotstr = np.array2string(self.rot, separator=', ').replace('\n', '')
        return f"CoordSys({self.origin!r}, array({rotstr}))"
